compiler/rustc_mir_dataflow/src/value_analysis.rs

```diff
@@ -0,0 +1,838 @@
+//! This module provides a framework on top of the normal MIR dataflow framework to simplify the
+//! implementation of analyses that track the values stored in places of interest.
+//!
+//! The default methods of [`ValueAnalysis`] (prefixed with `super_` instead of `handle_`)
+//! provide some behavior that should be valid for all abstract domains that are based only on the
+//! value stored in a certain place. On top of these default rules, an implementation should
+//! override some of the `handle_` methods. For an example, see `ConstAnalysis`.
+//!
+//! An implementation must also provide a [`Map`]. Before the anaylsis begins, all places that
+//! should be tracked during the analysis must be registered. Currently, the projections of these
+//! places may only contain derefs, fields and downcasts (otherwise registration fails). During the
+//! analysis, no new places can be registered.
+//!
+//! Note that if you want to track values behind references, you have to register the dereferenced
+//! place. For example: Assume `let x = (0, 0)` and that we want to propagate values from `x.0` and
+//! `x.1` also through the assignment `let y = &x`. In this case, we should register `x.0`, `x.1`,
+//! `(*y).0` and `(*y).1`.
+//!
+//!
+//! # Correctness
+//!
+//! Warning: This is a semi-formal attempt to argue for the correctness of this analysis. If you
+//! find any weak spots, let me know! Recommended reading: Abstract Interpretation. We will use the
+//! term "place" to refer to a place expression (like `mir::Place`), and we will call the
+//! underlying entity "object". For instance, `*_1` and `*_2` are not the same place, but depending
+//! on the value of `_1` and `_2`, they could refer to the same object. Also, the same place can
+//! refer to different objects during execution. If `_1` is reassigned, then `*_1` may refer to
+//! different objects before and after assignment. Additionally, when saying "access to a place",
+//! what we really mean is "access to an object denoted by arbitrary projections of that place".
+//!
+//! In the following, we will assume a constant propagation analysis. Our analysis is correct if
+//! every transfer function is correct. This is the case if for every pair (f, f#) and abstract
+//! state s, we have f(y(s)) <= y(f#(s)), where s is a mapping from tracked place to top, bottom or
+//! a constant. Since pointers (and mutable references) are not tracked, but can be used to change
+//! values in the concrete domain, f# must assume that all places that can be affected in this way
+//! for a given program point are already marked with top in s (otherwise many assignments and
+//! function calls would have no choice but to mark all tracked places with top). This leads us to
+//! an invariant: For all possible program points where there could possibly exist means of mutable
+//! access to a tracked place (in the concrete domain), this place must be assigned to top (in the
```

Comment:
Thank you for making this explicit!

Reply:
I think the wording can still be improved upon, but yes, thank you :)